contracts/extensions/MultisigPermissions.sol

```diff
@@ -0,0 +1,551 @@
+            motion.requiredPermissions == actionSummary.requiredPermissions,
+          "colony-multisig-invalid-motion"
+        );
+      }
+    }
+
+    validateMotionDomain(_permissionDomainId, _childSkillIndex, motionCount);
+    validateUserPermissions(_permissionDomainId, motionCount);
+
+    emit MotionCreated(msgSender(), motionCount);
+
+    changeVoteFunctionality(_permissionDomainId, motionCount, Vote.Approve, true);
+  }
+
+  function changeVote(
+    uint256 _permissionDomainId,
+    uint256 _childSkillIndex,
+    uint256 _motionId,
+    Vote _vote
+  ) public motionExists(_motionId) notExecuted(_motionId) notRejected(_motionId) {
+    validateMotionDomain(_permissionDomainId, _childSkillIndex, _motionId);
+    changeVoteFunctionality(_permissionDomainId, _motionId, Vote.Approve, _vote == Vote.Approve);
+    changeVoteFunctionality(_permissionDomainId, _motionId, Vote.Reject, _vote == Vote.Reject);
+
+    if (_vote == Vote.Reject && checkThreshold(_motionId, Vote.Reject)) {
+      cancelFunctionality(_motionId);
+    }
+  }
+
+  function changeVoteFunctionality(
+    uint256 _permissionDomainId,
+    uint256 _motionId,
+    Vote _vote,
+    bool _setTo
+  ) private {
+    if (_setTo) {
+      validateUserPermissions(_permissionDomainId, _motionId);
+    }
+
+    Motion storage motion = motions[_motionId];
+    bytes32 userPermissions = getUserRoles(msgSender(), _permissionDomainId);
+
+    uint8 roleIndex;
+    bool newlyAtThreshold;
+    bool anyBelowThreshold;
+
+    while (uint256(motion.requiredPermissions) >= (1 << roleIndex)) {
+      if ((motion.requiredPermissions & bytes32(1 << roleIndex)) > 0) {
+        // Then the motion requires this permission. Let's check it
+        // If the user is adding a vote but lacks the permission, skip
+        if (uint256(userPermissions) & (1 << roleIndex) == 0 && _setTo) {
+          roleIndex += 1;
+          continue;
+        }
+
+        // Update appropriately if vote changes
+        if (getUserVote(_motionId, msgSender(), roleIndex, _vote) != _setTo) {
+          setUserVote(_motionId, msgSender(), _vote, roleIndex, _setTo);
+
+          if (_setTo) {
+            motionVoteCount[_motionId][_vote][roleIndex] += 1;
+          } else {
+            motionVoteCount[_motionId][_vote][roleIndex] -= 1;
+          }
+
+          if (_vote == Vote.Approve) {
+            emit ApprovalChanged(msgSender(), _motionId, roleIndex, _setTo);
+          } else {
+            emit RejectionChanged(msgSender(), _motionId, roleIndex, _setTo);
+          }
+        }
+
+        if (_vote == Vote.Approve) {
+          uint256 threshold = getDomainSkillRoleThreshold(motion.domainSkillId, roleIndex);
+          if (motionVoteCount[_motionId][_vote][roleIndex] < threshold) {
+            anyBelowThreshold = true;
+          } else if (motionVoteCount[_motionId][_vote][roleIndex] == threshold && _setTo) {
+            newlyAtThreshold = true;
```

Comment:
Looks as though a user could delay failing execution indefinitely by continuing to refresh this timestamp.

Something like

```
motion.overallApprovalTimestamp  = (motion.overallApprovalTimestamp) ? motion.overallApprovalTimestamp : block.timestamp
```

seems like it would prevent that. I would have preferred using `min` or `max` but those wouldn't work here, since our goal is the intermediate value between `0` and `block.timestamp`.

Reply:
Also, perhaps we could rename that variable to `atThreshold` since the `newly` seems to be a bit wobbly :)

Or else -- can we make duplicate votes a no-op?

Reply:
I think with either solution it can't be prevented, as someone could withdraw their vote and then re-approve, which would (I would argue correctly) update the timestamp. Ultimately I don't think there's an issue here, but making duplicate votes a no-op in this context is straightforward so we probably should.